src/eva/vision/data/datasets/wsi.py

```diff
@@ -55,49 +60,34 @@ def _wsi(self) -> wsi.Wsi:
         wsi_obj.open_slide()
         return wsi_obj
 
-    def __getitem__(self, index: int) -> torch.Tensor:
-        # Calculate the desired zoom level based on target_mpp
-        level_idx, width, height = self._get_closest_level(self._wsi, self._target_mpp)
-
-        # Random Sampling
-        # TODO: make sampling method configurable
-        # TODO: add support for masking of unwanted regions
-        x_max, y_max = self._wsi.level_dimensions[level_idx]
-        x = random.randint(0, x_max - width)
-        y = random.randint(0, y_max - height)
+    @cached_property
+    def _coords(self) -> wsi.PatchCoordinates:
+        return wsi.PatchCoordinates.from_file(
+            wsi_path=self._file_path,
+            width=self._width,
+            height=self._height,
+            target_mpp=self._target_mpp,
+            backend=self._backend,
+            sampler=self._sampler,
+        )
 
+    @override
+    def __getitem__(self, index: int) -> torch.Tensor:
+        x, y = self._coords.x_y[index]
+        width, height, level_idx = self._coords.width, self._coords.height, self._coords.level_idx
         patch = self._wsi.read_region((x, y), (width, height), level_idx)
-        patch = self._apply_transforms(patch)
+        patch = self._apply_transforms(torch.from_numpy(patch))
```

Comment:
this seems to break the pipeline, since `torchvision.transforms.v2.ToImage()` permutes np.arrays but not tensors (see https://github.com/pytorch/vision/blob/main/torchvision/transforms/v2/functional/_type_conversion.py)

fixed in https://github.com/kaiko-ai/eva/pull/378

Reply:
Fixed, now. I ended up applying the dimension permute within the dataset class, to avoid typing issues with `_apply_transforms`.